src/bin/pgcopydb/ld_stream.c

```diff
@@ -767,8 +792,50 @@ streamRotateFile(LogicalStreamContext *context)
 		return true;
 	}
 
-	/* compute the WAL filename that would host the current LSN */
-	XLByteToSeg(context->cur_record_lsn, segno, context->WalSegSz);
+	/*
+	 * Determine the LSN to calculate walFileName in which to write the current
+	 * message.
+	 *
+	 * This walFileName calculation later ensures safe transaction formation in
+	 * the transform/apply process by always appending messages here to the
+	 * latest file and preventing rotation to earlier files.
+	 *
+	 * In most cases, jsonFileLSN should be the same as cur_record_lsn. However,
+	 * occasionally, current messages may have LSNs lower than the previous
+	 * ones. This can occur due to concurrent transactions with interleaved
+	 * LSNs. Since the logical decoding protocol sends the complete transaction
+	 * at commit time, the LSNs for messages within one transaction could be
+	 * lower than those of the previously streamed transactions. In such cases,
+	 * we use maximum LSN of the messages written so far to the disk in order to
+	 * write to the current file.
```

Comment:
I think we should copy/paste your ascii art drawing of 3 interleaved transactions here, it would make the whole thing so much easier to grasp when reading this part of the code.